[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/jalonso1979/puremacro/blob/main/curso/notebooks/T00_syllabus.ipynb)

> **Google Colab note**: the next cell installs `puremacro` and, if needed, fetches only the `curso/notebooks` folder (helpers, frozen data and models) from the public repository [jalonso1979/puremacro](https://github.com/jalonso1979/puremacro). It does not ask for Google Drive access.

In [1]:
# Environment setup: Google Colab, local install or Juno (iPad)
import os
import sys
from pathlib import Path

def _safe_probe(p, is_dir=None):
    """Does the path exist? Never raises OSError/PermissionError in the Juno sandbox."""
    try:
        p = Path(p)
        if is_dir is True:
            return p.is_dir()
        if is_dir is False:
            return p.is_file()
        return p.exists()
    except (OSError, PermissionError, TypeError, ValueError):
        return False

if "google.colab" in sys.modules:
    !pip install -q "puremacro>=4.0.1,<5" openpyxl
    if not _safe_probe(Path.cwd() / "_nbstyle.py", is_dir=False):
        # Partial clone of the public repository: only curso/notebooks, no history
        _repo = Path("/content/puremacro")
        if not _safe_probe(_repo, is_dir=True):
            !git clone -q --depth 1 --filter=blob:none --sparse https://github.com/jalonso1979/puremacro.git {_repo}
        !git -C {_repo} sparse-checkout set curso/notebooks 2>&1 | tail -n 2
        os.chdir(_repo / "curso/notebooks")

# Put the course helpers folder (_nbstyle, _datos) on sys.path, without leaving the sandbox
_candidates = [Path.cwd(), Path.cwd() / "notebooks", Path.cwd() / "curso/notebooks"]
try:
    _candidates.insert(0, Path(__file__).resolve().parent)  # Juno defines __file__
except (NameError, OSError):
    pass
for _c in _candidates:
    if _safe_probe(_c / "_nbstyle.py", is_dir=False):
        if str(_c) not in sys.path:
            sys.path.insert(0, str(_c))
        break

# Inline figures and the course style
%matplotlib inline
try:
    import _nbstyle
    _nbstyle.apply_style()
except ImportError:
    pass

# Advanced Macroeconomics (MAV) — Interactive Syllabus

## An Academic Manifesto: Modern Quantitative Macroeconomics

Modern macroeconomics has moved decisively beyond stylized pencil-and-paper abstractions. The representative-agent Robinson Crusoe economy—while historically vital for establishing intertemporal optimization concepts—cannot address the structural questions facing modern central banks, finance ministries, and multilateral institutions. 

Evaluating the macroeconomic impact of sudden stops, sovereign debt crises, climate policy transitions, or high-frequency monetary surprises requires models that satisfy three core requirements:

1. **Dynamic General Equilibrium Consistency**: Forward-looking households, firms, and policymakers optimize subject to budget and resource constraints, with equilibrium prices clearing goods, labor, and financial asset markets simultaneously.
2. **Microeconomic Heterogeneity and Incomplete Markets**: Wealth inequality, earnings risk, hand-to-mouth households, and credit frictions generate non-linear aggregate dynamics (as captured by Heterogeneous Agent New Keynesian, or HANK, frameworks) that representative-agent models systematically misstate.
3. **Rigorous Empirical Discipline**: Theoretical models are not mere mathematical curiosities; they are structural measurement instruments calibrated or estimated against national accounts, high-frequency financial surprises, and microeconomic survey microdata.

This course, **Advanced Macroeconomics (MAV)**, operationalizes this frontier. Built on **puremacro**—a pure Python computational suite running locally or in cloud Jupyter environments—each topic bridges dynamic mathematical theory, computational algorithms, and real-world empirical estimation.

## Methodological Architecture & Course Dynamics

In this course, models are computational laboratories. Every module is structured around an integrated four-stage research pipeline:

$$\text{Structural Economic Theory} \;\longrightarrow\; \text{Computational Formulation} \;\longrightarrow\; \text{Empirical Pipeline (Micro/Macro Data)} \;\longrightarrow\; \text{Policy Counterfactual Experiments}$$

- **Theory**: Formulate the dynamic optimization problem (Euler equations, Bellman equations, matching functions, or structural VAR identification restrictions).
- **Computational Solution**: Implement algorithms from value function iteration (VFI) to linearized perturbation methods, Kalman filtering, and local projections.
- **Empirical Measurement**: Confront models with real national accounts, FRED series, OECD vintages, and labor survey microdata.
- **Counterfactuals & Stress Tests**: Subject the model to historical stress tests (the 2008 Great Financial Crisis, the 2020 COVID shock, supply-chain dislocations, or sudden stops in emerging markets) to evaluate policy trade-offs.

In [2]:
import pandas as pd

modules = pd.DataFrame(
    [
        ("1", "Business Cycle Facts & Filtering", "Spectral Filters & High-Frequency Data", "T01_A, T01_B, T01_C"),
        ("2", "Neoclassical Growth & DSGE Local Solvers", "Dynare .mod, Klein QZ, RBC, New Keynesian, OccBin", "T02_A, T02_B, T02_C, T02_D"),
        ("3", "Risk, Uncertainty & AI (LLMs)", "GARCH, GAR, NLP Sentiment", "T03_A to T03_F"),
        ("4", "Real Business Cycles (RBC) & Dashboards", "Interactive Simulators", "T04_A, T04_B (Streamlit app)"),
        ("5", "Empirical Identification (Post-COVID Inflation)", "SVAR, Local Projections", "T05_A to T05_F"),
        ("6", "Incomplete Markets & HANK", "Heterogeneous Agents", "T06_A, T06_B, T06_C"),
        ("7", "Firm Dynamics & Goods Market", "Firm Dynamics & Market Power", "T07_A"),
        ("8", "Labor Flows & Frictions (ENOE Microdata)", "Search & Matching", "T08_A"),
        ("9", "Validation, Frontier & Custom Indexes", "Advanced Econometrics", "T09_A, T09_B"),
    ],
    columns=["Module", "Core Topic", "Tools / Concepts", "Active Labs"],
).set_index("Module")

# Display the curriculum architecture
modules

,Core Topic,Tools / Concepts,Active Labs
Module,,,
1,Business Cycle Facts & Filtering,Spectral Filters & High-Frequency Data,"T01_A, T01_B, T01_C"
2,Neoclassical Growth & DSGE Local Solvers,"Dynare .mod, Klein QZ, RBC, New Keynesian, OccBin","T02_A, T02_B, T02_C, T02_D"
3,"Risk, Uncertainty & AI (LLMs)","GARCH, GAR, NLP Sentiment",T03_A to T03_F
4,Real Business Cycles (RBC) & Dashboards,Interactive Simulators,"T04_A, T04_B (Streamlit app)"
5,Empirical Identification (Post-COVID Inflation),"SVAR, Local Projections",T05_A to T05_F
6,Incomplete Markets & HANK,Heterogeneous Agents,"T06_A, T06_B, T06_C"
7,Firm Dynamics & Goods Market,Firm Dynamics & Market Power,T07_A
8,Labor Flows & Frictions (ENOE Microdata),Search & Matching,T08_A
9,"Validation, Frontier & Custom Indexes",Advanced Econometrics,"T09_A, T09_B"


---

# Macroeconomic Experiment 0: Computational Onboarding Diagnostic Lab

### Economic Motivation: Trends, Non-Stationarity, and Shock Anatomy in Real-Time Data

Before entering the structural models of Topic 1, every quantitative macroeconomist must master the computational foundation of empirical macro: **data extraction, transformation, and stationarity diagnostics**.

In aggregate macroeconomic time series, raw national accounts data $Y_t$ (such as Real Gross Domestic Product) exhibit non-stationary stochastic trends with positive average drift:
$$Y_t = Y_0 \exp(\mu t + \xi_t), \quad \xi_t \sim I(1)$$
Standard linear regression and covariance-based spectral filtering (such as the Hodrick-Prescott filter or VAR estimation) require covariance-stationary inputs ($I(0)$). Applying these filters to raw non-stationary levels without understanding their integration order can produce spurious cyclicality and misleading correlation structures (the classic Slutsky-Yule effect).

Furthermore, macroeconomists measure growth rates at differing frequencies:
1. **Quarterly Annualized Growth Rate**:
   $$g_t^q = 400 \times \left( \log Y_t - \log Y_{t-1} \right)$$
   This measures the annualized pace of economic activity during a single quarter, compounding short-run volatility.
2. **Year-over-Year Annual Growth Rate**:
   $$g_t^{yoy} = 100 \times \left( \log Y_t - \log Y_{t-4} \right)$$
   This smooths seasonal idiosyncratic noise and captures medium-term momentum.

In this introductory diagnostic experiment, you will verify your local computational environment, evaluate the historical asymmetry between the **2008 Great Financial Crisis** and the **2020 COVID-19 pandemic shock**, and conduct formal **Augmented Dickey-Fuller (ADF)** stationarity tests to prove why modern macroeconomists must distinguish between $I(1)$ levels and $I(0)$ differences.

### Guided Student Tasks

1. **Pipeline Verification**: Load US quarterly Real GDP by extracting nominal GDP and the GDP deflator from `_datos.gasto()` (or directly from `GDPC1.csv` in `data_curso/`), constructing the chained volume series:
   $$Y_t = \frac{\text{GDP}_t^{\text{nominal}}}{\text{Deflator}_t} \times 100$$
2. **Growth Transformations**: Compute the log level ($100 \log Y_t$), annualized quarterly growth rate ($g_t^q$), and year-over-year rate ($g_t^{yoy}$).
3. **Shock Stress Test**: Isolate the maximum annualized quarterly contraction during the **2008 Great Financial Crisis** (2008Q1–2009Q4) versus the **2020 COVID-19 lockdown shock** (2020Q1–2020Q4). Compare how the annualized quarterly metric amplifies high-frequency shocks relative to year-over-year growth.
4. **Econometric Diagnostic (Unit Root Testing)**: Run the Augmented Dickey-Fuller test (Said & Dickey, 1984) testing the null hypothesis $H_0: \gamma = 0$ (unit root / non-stationarity) in:
   $$\Delta y_t = \alpha + \gamma y_{t-1} + \sum_{j=1}^p \delta_j \Delta y_{t-j} + \varepsilon_t$$
   The specification carries an intercept but no deterministic trend: that is `regression='c'`, the default in `adfuller` and what the code below runs.
   Compare the $p$-value of log Real GDP levels against quarterly growth rates.

In [3]:
import inspect
from types import SimpleNamespace

import numpy as np
import pandas as pd
from statsmodels.tsa.stattools import adfuller
import _datos

# ------------------------------------------------------------------------------
# 1. Pipeline Verification: Load real GDP and compute transformations
# ------------------------------------------------------------------------------
g = _datos.gasto()
pib_usa = g.xs("USA")["gdp"] / g.xs("USA")["gdp_defl"] * 100.0
log_pib = 100.0 * np.log(pib_usa)

# Quarterly annualized growth: 400 * Delta log(Y_t)
g_q = 400.0 * np.log(pib_usa).diff()

# Year-over-year growth: 100 * log(Y_t / Y_{t-4})
g_yoy = 100.0 * np.log(pib_usa / pib_usa.shift(4))

# ------------------------------------------------------------------------------
# 2. Historical Crisis Comparison: 2008 GFC vs. 2020 COVID-19 Shock
# ------------------------------------------------------------------------------
gfc_window = g_q.loc["2008-01-01":"2009-12-31"]
cov_window = g_q.loc["2020-01-01":"2020-12-31"]

gfc_min_q = gfc_window.min()
cov_min_q = cov_window.min()

gfc_trough_date = gfc_window.idxmin().to_period("Q")
cov_trough_date = cov_window.idxmin().to_period("Q")

gfc_min_yoy = g_yoy.loc["2008-01-01":"2009-12-31"].min()
cov_min_yoy = g_yoy.loc["2020-01-01":"2020-12-31"].min()

print("=" * 70)
print("MACROECONOMIC SHOCK ANATOMY: 2008 GFC vs 2020 COVID CONTRACTION")
print("=" * 70)
print(f"2008 GFC Trough (Annualized Quarterly): {gfc_min_q:.2f}% (Trough Quarter: {gfc_trough_date})")
print(f"2008 GFC Trough (Year-over-Year):      {gfc_min_yoy:.2f}%")
print(f"2020 COVID Trough (Annualized Quarterly): {cov_min_q:.2f}% (Trough Quarter: {cov_trough_date})")
print(f"2020 COVID Trough (Year-over-Year):       {cov_min_yoy:.2f}%")
print("-" * 70)

# ------------------------------------------------------------------------------
# 3. Econometric Diagnostic: Augmented Dickey-Fuller Stationarity Test
# ------------------------------------------------------------------------------
# statsmodels 0.15 returns a result object when called with
# ``result_object=True``, and warns about the legacy tuple otherwise; 0.14 --
# still what Colab serves -- does not take that keyword at all and raises
# TypeError.  Read the signature once and give both versions the same two
# attributes, so the notebook opens on either.
_ADF_OBJETO = "result_object" in inspect.signature(adfuller).parameters


def adf(serie):
    """ADF test exposing ``.statistic`` and ``.pvalue`` on statsmodels 0.14 and 0.15."""
    if _ADF_OBJETO:
        return adfuller(serie, result_object=True)
    estadistico, p_valor = adfuller(serie)[:2]
    return SimpleNamespace(statistic=estadistico, pvalue=p_valor)


adf_level = adf(log_pib.dropna())
p_val_level = adf_level.pvalue

adf_diff = adf(g_q.dropna())
p_val_diff = adf_diff.pvalue

print("AUGMENTED DICKEY-FULLER (ADF) TEST RESULTS:")
print(f"Log Real GDP Level:   Test Stat = {adf_level.statistic:.3f}, p-value = {p_val_level:.4f} -> {'Unit Root (Non-Stationary)' if p_val_level > 0.05 else 'Stationary'}")
print(f"Quarterly Growth (g): Test Stat = {adf_diff.statistic:.3f}, p-value = {p_val_diff:.4e} -> {'Unit Root' if p_val_diff > 0.05 else 'Stationary (I(0))'}")
print("=" * 70)

MACROECONOMIC SHOCK ANATOMY: 2008 GFC vs 2020 COVID CONTRACTION
2008 GFC Trough (Annualized Quarterly): -8.85% (Trough Quarter: 2008Q4)
2008 GFC Trough (Year-over-Year):      -4.06%
2020 COVID Trough (Annualized Quarterly): -32.82% (Trough Quarter: 2020Q2)
2020 COVID Trough (Year-over-Year):       -7.69%
----------------------------------------------------------------------
AUGMENTED DICKEY-FULLER (ADF) TEST RESULTS:
Log Real GDP Level:   Test Stat = -1.396, p-value = 0.5841 -> Unit Root (Non-Stationary)
Quarterly Growth (g): Test Stat = -12.935, p-value = 3.6295e-24 -> Stationary (I(0))


In [4]:
# ------------------------------------------------------------------------------
# Verification & Solution Validation
# ------------------------------------------------------------------------------
# Verify shock magnitudes match historical national accounting data
assert gfc_min_q < -8.0, f"Expected GFC annualized contraction < -8%, got {gfc_min_q:.2f}%"
assert cov_min_q < -30.0, f"Expected COVID annualized contraction < -30%, got {cov_min_q:.2f}%"
assert gfc_min_yoy < -3.5, f"Expected GFC YoY contraction < -3.5%, got {gfc_min_yoy:.2f}%"
assert cov_min_yoy < -7.0, f"Expected COVID YoY contraction < -7%, got {cov_min_yoy:.2f}%"

# Verify stationarity diagnostics
assert p_val_level > 0.10, f"Expected log GDP level p-value > 0.10 (unit root), got {p_val_level:.4f}"
assert p_val_diff < 0.01, f"Expected growth rate p-value < 0.01 (covariance stationary), got {p_val_diff:.4e}"

print("✓ Experiment 0 Diagnostic Check Passed: Environment, pipeline, and econometric properties verified successfully!")

✓ Experiment 0 Diagnostic Check Passed: Environment, pipeline, and econometric properties verified successfully!


### Economic Interpretation and Bridge to Topic 1

The empirical findings from this onboarding diagnostic establish two indispensable insights for the quantitative macroeconomist:

1. **Shock Profile and Asymmetry**: The 2008 Great Financial Crisis was a prolonged balance-sheet recession triggered by financial sector distress and collateral devaluation; annualized quarterly output contracted by $-8.85\%$ at its worst, but year-over-year output declined by $-4.06\%$. In contrast, the 2020 COVID-19 lockdown was a sudden, violent cessation of economic activity: annualized quarterly growth plunged to $-32.82\%$, an extreme tail shock nearly four times larger than the GFC trough. Yet because the shock was concentrated in 2020Q2, the year-over-year drop was $-7.69\%$.
2. **Integration Order and Econometric Validity**: The ADF test fails to reject the null hypothesis of a unit root for log Real GDP ($p = 0.5841$), confirming that macroeconomic levels are integrated of order 1 ($I(1)$). Attempting to estimate models or evaluate business cycle covariances directly on raw levels violates asymptotic normality and leads to spurious regressions. In contrast, log differences are unambiguously covariance-stationary ($p = 3.63 \times 10^{-24} \ll 0.01$).

With your computational pipeline and econometric foundations validated, proceed to **Module 1** (`T01_A_business_cycle_facts`) to decompose national accounts from the labor side and examine how filtering choices shape our understanding of economic fluctuations.